api/upload.py

```diff
@@ -166,6 +174,33 @@ def process_upload(request, strategy, container_type=None, id_=None, origin=None
 
 class Upload(base.RequestHandler):
 
+    def _create_ticket(self):
+        payload = self.request.json_body
+        metadata = payload.get('metadata', None)
+        filename = payload.get('filename', None)
+
+        if not (metadata or filename):
+            self.abort(404, 'metadata and filename are required')
+
+        tempdir = str(uuid.uuid4())
+        # Upload into a temp folder, so we will be able to cleanup
+        signed_url = files.get_signed_url(fs.path.join('tmp', tempdir, filename), config.fs, purpose='upload')
+
+        if not signed_url:
+            self.abort(405, 'Signed URLs are not supported with the current storage backend')
+
+        ticket = util.upload_ticket(self.request.client_addr, self.origin, tempdir, filename, metadata)
+        return {'ticket': config.db.uploads.insert_one(ticket).inserted_id,
+                'upload_url': signed_url}
+
+    def _check_ticket(self, ticket_id):
+        ticket = config.db.uploads.find_one({'_id': ticket_id})
+        if not ticket:
+            self.abort(404, 'no such ticket')
+        if ticket['ip'] != self.request.client_addr:
+            self.abort(400, 'ticket not for this resource or source IP')
```

Comment:
You could argue this is a 403 because ip is our rough permission check (other than ownership of the ticket).